orttraining/orttraining/core/graph/gradient_builder.cc

```diff
@@ -1101,6 +1101,38 @@ IMPLEMENT_GRADIENT_BUILDER(GetReduceLogSumExpGradient) {
   return result;
 }
 
+IMPLEMENT_GRADIENT_BUILDER(GetReduceL2Gradient) {
+  std::vector<NodeDef> result;
+  auto attributes = SrcNodeAttributes();
+  bool keepdims = true;
+  if (attributes.find("keepdims") != attributes.end() &&
+      attributes.at("keepdims").has_i()) {
+    keepdims = static_cast<bool>(attributes.at("keepdims").i());
+  }
+
+  ArgDef y_arg_def = O(0);
+  ArgDef dy_arg_def = GO(0);
+  if (!keepdims && attributes.find("axes") != attributes.end()) {
+    std::vector<int64_t> axes_values = RetrieveValues<int64_t>(attributes.at("axes"));
+    y_arg_def = IA("Unsqueezed_Y");
+    dy_arg_def = IA("Unsqueezed_dY");
+    result.emplace_back(NodeDef("Unsqueeze", {O(0)}, {y_arg_def}, {MakeAttribute("axes", axes_values)}));
+    result.emplace_back(NodeDef("Unsqueeze", {GO(0)}, {dy_arg_def}, {MakeAttribute("axes", axes_values)}));
+  }
+
+  result.emplace_back(NodeDef("Div", {dy_arg_def, y_arg_def}, {IA("Scaled_dY")}));
+
+  // Handle 0 elements in Y.
+  NodeDef zero_constant_node = ZeroConstantNode(IElemType(0));
+  ArgDef ZERO = zero_constant_node.output_args[0];
+  result.push_back(zero_constant_node);
+  result.emplace_back(NodeDef("Equal", {y_arg_def, ZERO}, {IA("Masked_Y")}));
+  result.emplace_back(NodeDef("Where", {IA("Masked_Y"), ZERO, IA("Scaled_dY")}, {IA("Masked_Scaled_dY")}));
```

Comment:
Cast might be faster then Where op here.

Reply:
When y=0, the scaled_dy=NaN, scaled_dy * cast(mask) on these elements is NaN * 0, which is not 0 I think.

Reply:
good point.